openedx/core/djangoapps/credentials_service/models.py

```diff
@@ -1,3 +1,264 @@
+    """
+
+    return 'certificate_template_assets/{id}/{filename}'.format(id=instance.id, filename=filename)
+
+
+def signatory_assets_path(instance, filename):
+    """
+    Custom path for signatories file assets.
+
+    Arguments:
+        instance: Signatory object
+        filename: file to upload
+
+    Returns:
+        Path of asset file e.g.
+        signatories/1/filename
+    """
+    return 'signatories/{id}/{filename}'.format(id=instance.id, filename=filename)
+
+
+def validate_image(image):
+    """
+    Validates that a particular image is small enough.
+    """
+    if image.size > (250 * 1024):
+        raise ValidationError(_('The image file size must be less than 250KB.'))
+
+
+class AbstractCredential(TimeStampedModel):
+    """
+    Abstract Credentials configuration model.
+    """
+    site = models.ForeignKey(Site)
+    is_active = models.BooleanField(default=False)
+
+    class Meta(object):
+        abstract = True
+
+
+class Signatory(TimeStampedModel):
+    """
+    Signatory model to add certificate signatories.
+    """
+    name = models.CharField(max_length=255)
+    title = models.CharField(max_length=255)
+    image = models.ImageField(
+        help_text=_(
+            'Image must be square PNG files. The file size should be under 250KB.'
+        ),
+        upload_to=signatory_assets_path,
+        validators=[validate_image]
+    )
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return '{name}, {title}'.format(
+            name=self.name,
+            title=self.title,
+        )
+
+    def save(self, *args, **kwargs):
+        """save the Signatory."""
+        if self.pk is None:
+            temp_image = self.image
+            self.image = None
+            super(Signatory, self).save(*args, **kwargs)
+            self.image = temp_image
+        else:
+            super(Signatory, self).save(*args, **kwargs)
+
+
+class CertificateTemplate(TimeStampedModel):
+    """
+    Certificate Template model to organize content for certificates.
+    """
+
+    name = models.CharField(max_length=255, db_index=True)
+    content = models.TextField(
+        help_text=_('HTML Template content data.')
+    )
+    certificate_type = models.CharField(
+        max_length=255, null=True, blank=True,
+        choices=MODE_CHOICES,
+    )
+    organization_id = models.IntegerField(
+        null=True,
+        blank=True,
+        db_index=True,
+        help_text=_('Organization with which this template is attached.'),
+    )
+
+    def __unicode__(self):
+        """Unicode representation. """
+        return '{name}'.format(
+            name=self.name
+        )
+
+
+class AbstractCertificate(AbstractCredential):
+    """
+    Abstract Certificate configuration to support multiple type of certificates
+    i.e. Programs, Courses.
+    """
+    signatories = models.ManyToManyField(Signatory)
+    template = models.ManyToManyField(CertificateTemplate, null=True, blank=True)
```

Comment:
Why is this many-to-many? Under what circumstances would a certificate have multiple templates?